packages/core/src/lib/normalize.unit.test.ts

```diff
@@ -0,0 +1,112 @@
+import { join } from 'node:path';
+import { describe, expect } from 'vitest';
+import { AuditOutput, Issue } from '@code-pushup/models';
+import { normalizeAuditOutputs, normalizeIssue } from './normalize';
+
+vi.mock('@code-pushup/utils', async () => {
+  const actual = await vi.importActual('@code-pushup/utils');
+
+  return {
+    ...actual,
+    gitRoot: vi.fn().mockResolvedValue('/User/code-pushup'),
+  };
+});
+
+describe('normalizeAuditOutputs', () => {
+  it('should forward empty auditOutputs', async () => {
+    await expect(normalizeAuditOutputs([])).resolves.toEqual([]);
+  });
+
+  it('should forward audits without details', async () => {
+    await expect(
+      normalizeAuditOutputs([{ slug: 'no-any' } as AuditOutput]),
+    ).resolves.toEqual([{ slug: 'no-any' }]);
+  });
+
+  it('should forward audit details with table', async () => {
+    const outputs = await normalizeAuditOutputs([
+      { details: { table: {} } } as unknown as AuditOutput,
+    ]);
+    expect(outputs.at(0)).toStrictEqual({ details: { table: {} } });
+  });
+
+  it('should forward audit details without issues', async () => {
+    const outputs = await normalizeAuditOutputs([
+      { details: { issues: undefined } } as unknown as AuditOutput,
+    ]);
+    expect(outputs.at(0)).toStrictEqual({ details: { issues: undefined } });
+  });
+
+  it('should forward audit details with empty issues', async () => {
+    const outputs = await normalizeAuditOutputs([
+      { details: { issues: [] } } as unknown as AuditOutput,
+    ]);
+    expect(outputs.at(0)).toStrictEqual({ details: { issues: [] } });
+  });
+
+  it('should forward audit details with issues and all undefined source', async () => {
+    const issues = [
+      { source: undefined },
+      { source: undefined },
+      { source: undefined },
+    ];
+    const outputs = await normalizeAuditOutputs([
+      { details: { issues } } as unknown as AuditOutput,
+    ]);
+    expect(outputs.at(0)?.details?.issues).toBe(issues);
+  });
+
+  it('should clone audit details with issues that have source specified', async () => {
+    const issues = [
+      { source: undefined },
+      { source: { file: 'index.js' } },
+      { source: undefined },
+    ] as Issue[];
+    await expect(
+      normalizeAuditOutputs([
+        { details: { issues } } as unknown as AuditOutput,
+      ]),
+    ).resolves.toStrictEqual([
+      {
+        details: {
+          issues: [
+            { source: undefined },
+            { source: { file: 'index.js' } },
+            { source: undefined },
+          ],
+        },
+      },
+    ]);
+  });
+});
```

Comment:
The "should forward" tests don't check that the audit object isn't cloned unnecessarily. To check the reference, use `.toBe()` instead of `.toStrictEqual()`.

Also, please stop using `as unknown as T` :pray: Just because we don't happen to have a lint rule for it, doesn't mean it doesn't have all the same problems of other unsafe typing patterns (`any`, etc.). There's always a better way. I removed `as unknown` in this file and found out that only 1 of 5 usages was actually suppressing a type error, and the 1 error was easily resolved with `[] as Issue[]` (without additional context, TS infers `[]` expressions as `never[]`, which then causes problems further down the line).

```suggestion
describe('normalizeAuditOutputs', () => {
  it('should forward empty auditOutputs', async () => {
    await expect(normalizeAuditOutputs([])).resolves.toEqual([]);
  });

  it('should forward audits without details', async () => {
    await expect(
      normalizeAuditOutputs([{ slug: 'no-any' } as AuditOutput]),
    ).resolves.toEqual([{ slug: 'no-any' }]);
  });

  it('should forward audit details with table', async () => {
    const audit = { details: { table: {} } } as AuditOutput;
    const outputs = await normalizeAuditOutputs([audit]);
    expect(outputs.at(0)).toBe(audit);
  });

  it('should forward audit details without issues', async () => {
    const audit = { details: { issues: undefined } } as AuditOutput;
    const outputs = await normalizeAuditOutputs([audit]);
    expect(outputs.at(0)).toBe(audit);
  });

  it('should forward audit details with empty issues', async () => {
    const audit = { details: { issues: [] as Issue[] } } as AuditOutput;
    const outputs = await normalizeAuditOutputs([audit]);
    expect(outputs.at(0)).toBe(audit);
  });

  it('should forward audit details with issues and all undefined source', async () => {
    const audit = {
      details: {
        issues: [
          { source: undefined },
          { source: undefined },
          { source: undefined },
        ],
      },
    } as AuditOutput;
    const outputs = await normalizeAuditOutputs([audit]);
    expect(outputs.at(0)?.details?.issues).toBe(audit.details?.issues);
  });

  it('should clone audit details with issues that have source specified', async () => {
    const issues = [
      { source: undefined },
      { source: { file: 'index.js' } },
      { source: undefined },
    ] as Issue[];
    await expect(
      normalizeAuditOutputs([{ details: { issues } } as AuditOutput]),
    ).resolves.toStrictEqual([
      {
        details: {
          issues: [
            { source: undefined },
            { source: { file: 'index.js' } },
            { source: undefined },
          ],
        },
      },
    ]);
  });
});
```